Sources/SotoS3FileTransfer/S3FileTransferManager.swift

```diff
@@ -412,11 +413,12 @@ public class S3FileTransferManager {
                 }
                 // construct list of files to delete, if we are doing deletion
                 if delete == true {
+                    let targetKeys = Set(targetFiles.map { $0.to.key })
                     let deletions = s3Files.compactMap { s3File -> S3File? in
-                        if targetFiles.first(where: { $0.to.key == s3File.file.key }) == nil {
-                            return s3File.file
-                        } else {
+                        if targetKeys.contains(s3File.file.key) {
                             return nil
+                        } else {
+                            return s3File.file
```

Comment:
I've flipped the condition to avoid double negation (`if !targetKeys.contains(...) { ... } else { ... }` but I can switch that back if you prefer to keep it closer to the original.

Reply:
Or maybe

```swift
                    let deletions = s3Files
                        .filter { !targetKeys.contains($0.file.key) }
                        .map(\.file)
```
?

Reply:
You're going through the array twice, better to use one `compactMap`.